inference-engine/src/mkldnn_plugin/nodes/mkldnn_shuffle_channels_node.cpp

```diff
@@ -0,0 +1,338 @@
+        supportedPrimitiveDescriptors.push_back({config, impl_type, memoryFormat});
+    };
+
+    if (mayiuse(cpu::x64::sse41)) {
+        if (nDims == 4) {
+            pushSupportedPrimitiveDescriptor(mkldnn::memory::format_tag::nhwc);
+            if (axis != 1) {
+                if (impl_desc_type::jit_avx512 == impl_type) {
+                    pushSupportedPrimitiveDescriptor(mkldnn::memory::format_tag::nChw16c);
+                } else {
+                    pushSupportedPrimitiveDescriptor(mkldnn::memory::format_tag::nChw8c);
+                }
+            }
+        } else if (nDims == 5) {
+            pushSupportedPrimitiveDescriptor(mkldnn::memory::format_tag::ndhwc);
+            if (axis != 1) {
+                if (impl_desc_type::jit_avx512 == impl_type) {
+                    pushSupportedPrimitiveDescriptor(mkldnn::memory::format_tag::nCdhw16c);
+                } else {
+                    pushSupportedPrimitiveDescriptor(mkldnn::memory::format_tag::nCdhw8c);
+                }
+            }
+        }
+    }
+
+    pushSupportedPrimitiveDescriptor(MKLDNNMemory::GetPlainFormat(getParentEdgeAt(0)->getDims()));
+}
+
+void MKLDNNShuffleChannelsNode::createPrimitive() {
+    if (prim || !mayiuse(cpu::x64::sse41))
+        return;
+    auto &dstMemPtr = getChildEdgeAt(0)->getMemoryPtr();
+    auto &srcMemPtr = getParentEdgeAt(0)->getMemoryPtr();
+    if (!dstMemPtr || !dstMemPtr->GetPrimitivePtr())
+        THROW_SHCH_ERROR << "has not allocated destination memory";
+    if (!srcMemPtr || !srcMemPtr->GetPrimitivePtr())
+        THROW_SHCH_ERROR << "has not allocated input memory";
+    if (getSelectedPrimitiveDescriptor() == nullptr)
+        THROW_SHCH_ERROR << "has unidentified preferable primitive descriptor";
+
+    const bool isBlocked = getParentEdgeAt(0)->getMemory().GetDesc().isBlockedCFormat();
+
+    int batchRank = axis;
+    int sptialRank = dataRank - axis - 1;
+
+    // 2 for decompose axis dim, 1 for composed spatial dim
+    int reshapedRank = batchRank + 2 + static_cast<int>(sptialRank != 0) + static_cast<int>(isBlocked && (sptialRank == 0));
+    order.resize(reshapedRank);
+    optimizedParams.src_block_dims.resize(reshapedRank);
+
+    size_t spatialShapeSize = 1;
+    if (sptialRank != 0) {
+        for (int i = batchRank + 1; i < dataRank; i++) {
+            spatialShapeSize *= dataDims[i];
+        }
+    }
+
+    const int channelDim = 1;
+    if (isBlocked) {
+        size_t blkSize = mayiuse(cpu::x64::avx512_common) ? 16 : 8;
+        size_t CB = div_up(dataDims[1], blkSize);
+        SizeVector srcBlockedDims = getParentEdgeAt(0)->getDesc().getBlockingDesc().getBlockDims();
+        if (axis > channelDim) {  // axis on spatial
+            for (int i = 0; i < batchRank; i++) {
+                order[i] = i;
+                optimizedParams.src_block_dims[i] = srcBlockedDims[i];
+            }
+            optimizedParams.src_block_dims[batchRank] = group;
```

Comment:
This code (lines: 161 - 167) are used three times (for each layout format): lines 161, 204, 227. May be should create lambda function with this code?

Reply:
Yes, that's more clean and clear. Created lambda decomposeAndTranspose() and call this anywhere possible. Thank you!